### PR TITLE
Added CPU offloading

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Added CPU offloading. Compilation takes no more than 1x GPU memory. Before engine compilation, the model and graph module are moved to CPU.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 1]
```diff
@@ -733,7 +731,11 @@ def run(
         self._create_timing_cache(
             builder_config, self.compilation_settings.timing_cache_path
         )
-
+        # TODO: Memory control prototyping. Under discussion
+        if self.compilation_settings.offload_module_to_cpu:
+            del self.module
+            gc.collect()
+            torch.cuda.empty_cache()
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:738]: use the delete_module function here

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 2]
```diff
@@ -488,14 +488,12 @@ def _save_weight_mapping(self) -> None:
...
-                for k, v in self.module.state_dict().items()
-            }
-        else:
-            sd = {k: v.reshape(-1) for k, v in self.module.state_dict().items()}
+        # If the model original position is on CPU, move it GPU
+        sd = {
+            k: v.reshape(-1).to(torch_device)
+            for k, v in self.module.state_dict().items()
+        }
+
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:496]: Is `gm_is_on_cuda` still necessary?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 3]
```diff
@@ -684,6 +677,9 @@ def compile(
     )
 
     gm = exported_program.module()
+    # TODO: Memory control prototyping. Under discussion
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/_compiler.py:680]: Is the PR good for review or still in progress?

Comment (cehongwang) [reply py/torch_tensorrt/dynamo/_compiler.py]: Good for review

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py (hunk 2 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: No. For the new version of compilation, part of the weights are offloaded to CPU for some reason. So here we need to move all weights to GPU.
https://github.com/pytorch/TensorRT/issues/3450

**Review by narendasan — COMMENTED** [review]
Wasnt there supposed to be a bunch of logging?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 4]
```diff
@@ -684,12 +678,17 @@ def compile(
...
 
     # Apply lowering on the graph module
     gm = post_lowering(gm, settings)
     logger.debug("Lowered Input graph: " + str(gm.graph))
-
+    if offload_module_to_cpu:
+        exported_program.module().to(CPU_DEVICE)
+        logger.info(
+            "The model is offloaded to CPU during compilation. If you want to keep the model on GPU, set offload_module_to_cpu=False."
+        )
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_compiler.py:691]: Consider this message: The PyTorch model was moved to the CPU to allocate all GPU memory to TensorRT. To retain the model on the GPU, set offload_module_to_cpu=False

Comment (peri044) [reply py/torch_tensorrt/dynamo/_compiler.py:691]: Also, there's one more thing we discussed - throw a warning if we predict an oom when offload_module_to_cpu=False. This could be achieved by measuring the size of the pytorch module and available GPU memory.

**Review by peri044 — COMMENTED** [review]
We need to test this change across our entire test suite to ensure it is working as expected.

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 5]
```diff
@@ -833,6 +840,7 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
                 str(name),
                 str(submodule.graph),
             )
+            submodule.to(to_torch_device(settings.device))
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_compiler.py:843]: Can you add a test case with fallback to ensure this works as expected?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 6]
```diff
@@ -690,6 +684,18 @@ def compile(
...
+    if offload_module_to_cpu:
+        exported_program.module().to(CPU_DEVICE)
+        logger.info(
+            "The PyTorch model was moved to the CPU to allocate all GPU memory to TensorRT. To retain the model on the GPU, set offload_module_to_cpu=False"
+        )
+    else:
+        remaining_memory, total_memory = torch.cuda.mem_get_info()
+        if remaining_memory < total_memory // 2:
+            logger.warning(
+                "The remaining GPU memory is not enough to compile the model. This may lead to an OOM error. Consider setting offload_module_to_cpu=True."
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:697]: Maybe make this a bit more tentative: `Remaining GPU memory may not be enough to compile the TensorRT engine for this model resulting in an OOM error, Consider setting offload_module_to_cpu=True`

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 7]
```diff
@@ -690,6 +685,18 @@ def compile(
     gm = post_lowering(gm, settings)
     logger.debug("Lowered Input graph: " + str(gm.graph))
 
+    # Move the weights in the state_dict to CPU
+    if offload_module_to_cpu:
+        exported_program.module().to(CPU_DEVICE)
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_compiler.py:690]: I encountered a situation where this wasn't enough and it required calling torch.cuda.empty_cache and gc.collect as well to release the memory. Here's a suggestion: Modify the `delete_module` function to `deallocate_module(module, delete=False)` and call it here as `deallocate_module(exported_program.module(), delete=False)`.

**Review by peri044 — APPROVED** [review]
LGTM
